README.md
# Cubes

This package contains the Cubes library, for quick prototyping. It also contains a Storybook Web app.

---

- [Cubes](#cubes)
  - [Development Workflow](#development-workflow)
    - [Create a New Component Guidelines](#create-a-new-component-guidelines)
    - [Write Efficient Storybook Stories](#write-efficient-storybook-stories)
  - [Directory Structure](#directory-structure)
  - [Typescript Structure](#typescript-structure)
  - [Publish :rocket:](#publish-)

## Development Workflow

Simply start the Web Storybook and then develop components!

```shell
$ yarn start
```

### Create a New Component Guidelines

Let's say we want to develop a new component, named `MyComponent`.

1. Decide in which category your component should be (see `/src/components` subfolders).
2. Create a subfolder with your component's name, with a file in it with the same name. Following our example: `src/components/<category>/MyComponent/MyComponent.tsx`. The component **must** be exported using `export const MyComponent`. **NO** `export default`. Using `export default` would break Typescript Documentation generation.
3. Along with our new file, create a Storybook stories file: `MyComponent.stories.tsx`.
4. Reference your component in `src/components/index.ts`
5. Develop your component and look at it using Storybook. :rocket:

For additional guidelines on how to develop a component, see [Develop a New Component](/docs/mobile-app/Develop-a-New-Component).

### Write Efficient Storybook Stories

> **Disclaimer:** This section needs to be expanded after upgrading to Storybook 6. These guidelines are for Storybook 5 but will stay similar for the next major Storybook version.

[Storybook "Writing Stories" Documentation](https://storybook.js.org/docs/basics/writing-stories/) is a good starting point in case you don't know how to write stories, but here we expand on some guidelines we want to follow.

- Stories file name should have the name of the component, followed with `.stories.tsx`.
- Make sure to pass the concerned component in the `component` stories definition.
- The name of the story should have the same name as their component too, in order for Typescript Docgen parser to work correctly.
- Good component documentation starts with the JSDoc. Make sure to comment correctly every component along with all of their properties.
- The first story appears differently in Storybook "Docs" tab (only in Storybook Web). The first story should be the one representing what the component should do (its primary purpose). The first story could have a different format than the other ones to show more capabilities of the component, if needed.
- Stories should show usage of every component prop. Usage of multiple props could be shown in a single story, especially if the component has a lot of properties.
- If pertinent, some stories may not necessarily show a component but could instead show a composite component made of basic components, to better illustrate possible usages.
- Each story code should be "standalone". It should not be encapsulated in a function, as this would be irrelevant when looking at the source code of the story (see example below).

**Example Story:**

```tsx
import React from 'react'
import { MyComponent } from './MyComponent'
import { fileAbsolute } from 'paths.macro'
import { getStoryTitle } from '../../../storybook/get-story-title'

// Declaration of the Stories configuration for this component
export default {
  title: getStoryTitle(fileAbsolute), // Component Name (automatically generated from file name)
  component: MyComponent, // Component instance
}

// Declare stories using `export const`
// The first story is shown differently in "Docs" Tab.
// If the first story receives a props parameter, it will show controls in the Props Table to dynamically change props.
export const MyFirstStory: React.FC = props => <MyComponent {...props} />
export const MySecondStory: React.FC = () => <MyComponent loading />

// DON'T ENCAPSULATE COMPONENT USAGE IN FUNCTION LIKE SO:
const funcReturningMyComponent = () => <MyComponent prop="value">
export const MyNotSoCoolStory = funcReturningMyComponent()
```

> **Additional Notes**
>
> - There should be stories for every component of the DS, for documentation purpose.
> - For now, we only have stories written in Typescript, but later on, we could expand them further using MDX.

## Directory Structure

- `src/` contains sources for the package. Output does not contain the `src/storybook/` folder and the `*.stories.tsx` files.
- `storybook/` contains the storybook related code. `web/` subdirectory is for the Web Storybook.
- `public/` directory contains the basic React App assets for Storybook to work with.

## Typescript Structure

**VERY IMPORTANT NOTES**

There are 2 `tsconfig.json` files in this package:

- `tsconfig.json`: Used by VS Code for code completion and stuff and when building Storybook (doesn't emit anything).
- `tsconfig.build.json`: Used when building the package (output code to `dist/` folder).

## Publish :rocket:

Deploy the package to NPM using:

```shell
$ yarn version && yarn publish --access public
```


CHANGELOG.md
## <small>0.0.2-alpha.9 (2020-07-16)</small>

- Add Transparent variant for top bar. ([b5c14c6](https://github.com/Trampoline-CX/Cubes/commit/b5c14c6))
- Add BottomNavigationBar. ([aa1d3e0](https://github.com/Trampoline-CX/Cubes/commit/aa1d3e0))
- Add Chromatic (#5) ([90ce372](https://github.com/Trampoline-CX/Cubes/commit/90ce372)), closes [#5](https://github.com/Trampoline-CX/Cubes/issues/5)
- Add Chromatic automatic changes acceptance for protected branches. (#29) ([eb86a60](https://github.com/Trampoline-CX/Cubes/commit/eb86a60)), closes [#29](https://github.com/Trampoline-CX/Cubes/issues/29)
- Add comments and remove unused code. ([0d44298](https://github.com/Trampoline-CX/Cubes/commit/0d44298))
- Add ProgressBar Cube. (#12) ([aa0c3f0](https://github.com/Trampoline-CX/Cubes/commit/aa0c3f0)), closes [#12](https://github.com/Trampoline-CX/Cubes/issues/12)
- Add react-native-safe-area-context and Screen component. ([7a55ab8](https://github.com/Trampoline-CX/Cubes/commit/7a55ab8))
- Add right icon to basic story of TopBar. ([616643a](https://github.com/Trampoline-CX/Cubes/commit/616643a))
- Add Screen.Content and fix BottomNavigationBar. ([0814fde](https://github.com/Trampoline-CX/Cubes/commit/0814fde))
- Add Stack.reverse prop + Make it possible to add icons to the right of the TabBar. ([7b1fb08](https://github.com/Trampoline-CX/Cubes/commit/7b1fb08))
- Add TopBar and NavigationProvider components. Review components exports to have a barrel file for ea ([4a153ce](https://github.com/Trampoline-CX/Cubes/commit/4a153ce))
- Adjust timeout. (#27) ([af66206](https://github.com/Trampoline-CX/Cubes/commit/af66206)), closes [#27](https://github.com/Trampoline-CX/Cubes/issues/27)
- Attempt to fix Chromatic UI tests (#62) ([ef8bb82](https://github.com/Trampoline-CX/Cubes/commit/ef8bb82)), closes [#62](https://github.com/Trampoline-CX/Cubes/issues/62)
- Automatically Merge master in develop (#50) ([ff97c10](https://github.com/Trampoline-CX/Cubes/commit/ff97c10)), closes [#50](https://github.com/Trampoline-CX/Cubes/issues/50)
- Avatar (#10) ([1687ada](https://github.com/Trampoline-CX/Cubes/commit/1687ada)), closes [#10](https://github.com/Trampoline-CX/Cubes/issues/10)
- Basic Navigation (#1) ([5ba2fb3](https://github.com/Trampoline-CX/Cubes/commit/5ba2fb3)), closes [#1](https://github.com/Trampoline-CX/Cubes/issues/1)
- Change description of components to be more usage-aligned. (#20) ([fe2c805](https://github.com/Trampoline-CX/Cubes/commit/fe2c805)), closes [#20](https://github.com/Trampoline-CX/Cubes/issues/20)
- Change package name + version ([f57cda7](https://github.com/Trampoline-CX/Cubes/commit/f57cda7))
- Changed autoAcceptChanges to work only on develop. (#39) ([e634895](https://github.com/Trampoline-CX/Cubes/commit/e634895)), closes [#39](https://github.com/Trampoline-CX/Cubes/issues/39)
- Checkbox (#14) ([03d3b8c](https://github.com/Trampoline-CX/Cubes/commit/03d3b8c)), closes [#14](https://github.com/Trampoline-CX/Cubes/issues/14)
- ChoiceList (#21) ([d995f43](https://github.com/Trampoline-CX/Cubes/commit/d995f43)), closes [#21](https://github.com/Trampoline-CX/Cubes/issues/21)
- Chromatic Specific Stories Configuration (#46) ([0471370](https://github.com/Trampoline-CX/Cubes/commit/0471370)), closes [#46](https://github.com/Trampoline-CX/Cubes/issues/46)
- Complete AppProvider stories and JSDoc. ([fadb7f5](https://github.com/Trampoline-CX/Cubes/commit/fadb7f5))
- Dialog, Alert and Confirm (#9) ([75b0f5b](https://github.com/Trampoline-CX/Cubes/commit/75b0f5b)), closes [#9](https://github.com/Trampoline-CX/Cubes/issues/9)
- Empty State (#32) ([06d16fe](https://github.com/Trampoline-CX/Cubes/commit/06d16fe)), closes [#32](https://github.com/Trampoline-CX/Cubes/issues/32)
- Fix AppProvider duplicated props. (#15) ([83c4e07](https://github.com/Trampoline-CX/Cubes/commit/83c4e07)), closes [#15](https://github.com/Trampoline-CX/Cubes/issues/15)
- Fix BottomNavigationBar background on native. (#2) ([1331952](https://github.com/Trampoline-CX/Cubes/commit/1331952)), closes [#2](https://github.com/Trampoline-CX/Cubes/issues/2)
- Fix build + remove storybook react-native. ([03a68c7](https://github.com/Trampoline-CX/Cubes/commit/03a68c7))
- Fix dependency cycles. (#30) ([0ebf28e](https://github.com/Trampoline-CX/Cubes/commit/0ebf28e)), closes [#30](https://github.com/Trampoline-CX/Cubes/issues/30)
- Fix dialog on Native (#51) ([c3e01bd](https://github.com/Trampoline-CX/Cubes/commit/c3e01bd)), closes [#51](https://github.com/Trampoline-CX/Cubes/issues/51)
- Fix Eslint config + issues. ([4051ba1](https://github.com/Trampoline-CX/Cubes/commit/4051ba1))
- Fix Extract Release Version not getting executed. (#43) ([e333fd4](https://github.com/Trampoline-CX/Cubes/commit/e333fd4)), closes [#43](https://github.com/Trampoline-CX/Cubes/issues/43)
- Fix install issue. ([8189688](https://github.com/Trampoline-CX/Cubes/commit/8189688))
- GitHub Actions - Build, Lint and Deploy to Chromatic (#18) ([ff6267e](https://github.com/Trampoline-CX/Cubes/commit/ff6267e)), closes [#18](https://github.com/Trampoline-CX/Cubes/issues/18)
- Initial commit ([b6e028d](https://github.com/Trampoline-CX/Cubes/commit/b6e028d))
- Limit TopBar title size to be one line only. ([f98f532](https://github.com/Trampoline-CX/Cubes/commit/f98f532))
- List (#7) ([cc75cd9](https://github.com/Trampoline-CX/Cubes/commit/cc75cd9)), closes [#7](https://github.com/Trampoline-CX/Cubes/issues/7)
- ListView and ListView.Item (#6) ([44d5dc6](https://github.com/Trampoline-CX/Cubes/commit/44d5dc6)), closes [#6](https://github.com/Trampoline-CX/Cubes/issues/6)
- Make scrolling of Screen.Content an opt-out feature. ([f1ddf71](https://github.com/Trampoline-CX/Cubes/commit/f1ddf71))
- Material Icons (#26) ([7cda1c8](https://github.com/Trampoline-CX/Cubes/commit/7cda1c8)), closes [#26](https://github.com/Trampoline-CX/Cubes/issues/26)

storybook/web/main.ts
import path from 'path'
import type { StorybookConfig } from '@storybook/core/types'

const tsconfig = path.resolve(__dirname, '../../tsconfig.json')

const storybookConfig: StorybookConfig = {
  stories: ['../../src/components/**/*.stories.@(tsx|mdx)'],
  addons: [
    '@storybook/addon-actions',
    '@storybook/addon-toolbars',
    '@storybook/addon-docs',
    '@storybook/addon-controls',
  ],
  typescript: {
    check: true,
    checkOptions: { tsconfig },
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      tsconfigPath: tsconfig,
      propFilter: (prop: { name: string }): boolean => !/^(testID)$/.test(prop.name),
      compilerOptions: {
        allowSyntheticDefaultImports: true,
        esModuleInterop: true,
      },
    },
  },
  webpackFinal: async config => {
    if (!config?.resolve) {
      return config
    }

    if (config.module) {
      /**
       * Use metro-react-native-babel-preset for loading react-native-vector-icons, as otherwise
       * Storybook crashes at launch with following error:
       *
       * ERROR in ./node_modules/react-native-vector-icons/lib/create-icon-set.js 43:21
       * Module parse failed: Unexpected token (43:21)
       * You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
       * |
       * |   class Icon extends PureComponent {
       * >     static propTypes = {
       * |       allowFontScaling: PropTypes.bool,
       * |       name: IconNamePropType,
       *  @ ./node_modules/react-native-vector-icons/MaterialIcons.js 6:0-50 9:16-29
       */
      config.module.rules.push({
        test: /\.js$/,
        include: [
          path.resolve(__dirname, '../../node_modules/react-native-vector-icons/'),
          path.resolve(__dirname, '../../node_modules/@ptomasroos/react-native-multi-slider/'),
        ],
        use: {
          loader: 'babel-loader',
          options: {
            cacheDirectory: false,
            presets: [
              ['module:metro-react-native-babel-preset', { disableImportExportTransform: true }],
            ],
          },
        },
      })
    }

    config.resolve.alias = {
      ...(config.resolve.alias || {}),
      // Replace react-native dependencies with react-native-web
      'react-native$': 'react-native-web',
      // Replace @storybook/react-native with @storybook/react
      '@storybook/react-native': '@storybook/react',
      // Make react-native-svg work
      'react-native-svg': 'react-native-svg/lib/commonjs/ReactNativeSVG.web',
      // Mock react-native-safe-area-context
      'react-native-safe-area-context': path.resolve(
        __dirname,
        '../../src/storybook/utils/react-native-safe-area-context',
      ),
      // Mock react-native-screens
      'react-native-screens': path.resolve(
        __dirname,
        '../../src/storybook/utils/react-native-screens',
      ),
      // Mock expo-haptics
      'expo-haptics$': path.resolve(__dirname, '../../src/storybook/utils/expo-haptics'),
      // Mock @expo/vector-icons with react-native-vector-icons
      '@expo/vector-icons$': path.resolve(__dirname, '../../src/storybook/utils/expo-vector-icons'),

      // Correctly use `.web.js` files in our own code.
      './DatePickerView/DatePickerView': path.resolve(
        __dirname,
        '../../src/components/forms/DatePicker/DatePickerView/DatePickerView.web',
      ),
      './TimePickerView/TimePickerView': path.resolve(
        __dirname,
        '../../src/components/forms/TimePicker/TimePickerView/TimePickerView.web',
      ),
    }

    return config
  },
}

export default storybookConfig
